README
Copyright (C) 2008, 2009, 2010 Mihai Şucan

$URL: http://code.google.com/p/paintweb $
$Date: 2014-01-28 12:33:54 $

PaintWeb is a Web application which allows users to draw inside the Web browser 
making use of new Web technologies.

This folder holds the PaintWeb source code, documentation and the packaged 
PaintWeb build.


Requirements:

  A Web browser which implements the HTML 5 <canvas> 2D context API.
  Known to work are: Opera 9.5+, Firefox 3+, Safari 3.1+, Chrome 3+, Konqueror 
  4.1+, Microsoft Internet Explorer 9+.
  Each Web browser has its minor quirks.


Usage:

  You can use PaintWeb inside Moodle, or you may integrate PaintWeb into your 
  project via TinyMCE. The "ext" folder holds a TinyMCE plugin which integrates 
  PaintWeb into TinyMCE. Please check the documentation for further details:

  http://code.google.com/p/paintweb

Installation:

  If you want to learn how to repackage PaintWeb yourself, please see the 
  INSTALL file.

Files and folders in this package:

  build/   - holds the packaged PaintWeb build.
  CHANGES  - holds a chronologically ordered list of changes.
  config-default.mk - the default Makefile configuration.
  LICENSE  - the BSD license.
  demos/   - holds usage demos with PaintWeb.
  docs/    - holds project documentation.
  ext/     - holds code for integrating PaintWeb into external projects. For
             example, there's code for Moodle and TinyMCE integration.
  INSTALL  - PaintWeb packaging instructions.
  Makefile - the Makefile script which drives the PaintWeb packaging process.
  README   - this file
  scripts/ - holds scripts used for PaintWeb packaging.
  src/     - holds the PaintWeb source code.
  tests/   - holds various tests made during the development of PaintWeb.

Get in contact:

  - You can ask questions about the project in the public mailing list of the 
  project. You can post your message by sending an email to 
  paintweb@googlegroups.com.
  - You can contact me via email at mihai.sucan@gmail.com or via my site 
  www.robodesign.ro/mihai. Any feedback is always welcome.
  - I am always looking for volunteers to help me with coding - there's lots of 
  work to do. We can collaborate via Gtalk, Yahoo, IRC, ICQ and anyhow you like.
  - On IRC I'm online everyday on irc.freenode.net. Just /query ROBOd - my 
  nickname.

Thank you!


-- 
vim:set spell spl=en fenc=utf-8 ff=unix: 



CHANGES
2014-01-28:
===========

- Switch the code to the BSD license.

2010-06-26:
===========

- Implemented support for Microsoft Internet Explorer 9, technical preview 3.
- Made the code to not rely on the existence of the global Node object.
- Made the code to not rely on the expression evaluation:
  if (window.getComputedStyle)
  because MSIE 9 reports that as false (window.getComputedStyle is null), even 
  if it does actually implement support for the method. This is similar to how 
  document.all is hidden in Opera.
- Made the code to not rely on the successful execution of elem.focus(). MSIE 9 
shows an error when code tries to focus an invisible element.
- MSIE 9 does not properly parse the user interface markup when it is loaded 
using the XMLHttpRequest. This broke the PaintWeb initialization when it loaded 
from the source code, without being packaged. I worked around the issue by using 
the xhr.responseText, instead of the responseXML result.
- Fixed a bug in the pwlib.isSameHost() method which prevented it from working 
properly in MSIE 9.

Known issues:
- MSIE 9 does not yet implement the globalCompositeOperation property, which 
makes the Color mixer visualisation look weird.
- Changing elem.style.width/height causes the Canvas element to be cleared, 
  which is entirely wrong. This breaks:
  - The PaintWeb.imageCrop() method that is used by the image resize handle (the 
      bottom right corner). I was able to fix this by reordering the code.
  - The PaintWeb.updateCanvasScaling() method. This issue cannot be worked 
  around and it causes PaintWeb to display a transparent image on load.
  - The PaintWeb.imageLoad() method. This issue cannot be worked around - at the 
  moment - and it causes for any image loaded by PaintWeb to disappear ... erm, 
  in the same method execution. Effectively, you cannot see the image you loaded 
  into PaintWeb.
  - The PaintWeb.imageZoomTo() method. This issue cannot be worked around. Image 
  zoom relies on having the Canvas element scaled up with CSS.
- TinyMCE 3.2.5 does not work in MSIE 9.

All in all, thank you Microsoft!

2009-11-16:
===========

- New English strings for Moodle 1.9.

2009-11-10:
===========

- Another change for the Color Bucket tool. Now the getImageData() method is 
invoked for each line. This is a compromise between lots of memory usage and 
performance.

2009-11-09:
===========

- Changed a bit the way the new Color Bucket tool works. The getImageData() 
  method is very slow on Gecko. Now flood fill runs good enough on Gecko as 
  well.

2009-11-08:
===========

- Implemented the new Color Bucket tool. This fixes issue #2 :)
- Updated the default PaintWeb configuration to include the new color bucket 
tool. Updated the Moodle configuration file as well.
- Updated the default Makefile configuration to include the new color bucket 
tool.
- Fixed a bug in the interface script - it caused failure when registering tools 
which were not in the interface layout.
- Using the new mozImageSmoothingEnabled added in Gecko 1.9.2 for the Canvas 2D 
context. This disables the smooth scaling of the Canvas element on the OLPC XO.
- Updated the English localization file to include the new strings for the color 
bucket tool.

2009-11-06:
===========

- Improvements to the imagesave19.php script used for saving images in Moodle 
1.9.

2009-11-04:
===========

- Fixed a bug in the TinyMCE plugin. The overlay button did not always show in 
the editor. This bug affected Moodle.

2009-10-29:
===========

- Work on Moodle 1.9 integration. Made improvements to the Moodle extension for 
PaintWeb and fixed the imagesave19.php script.

2009-10-10:
===========

- Fixed issue #33 by adding a configuration file for the Makefile. Invoke make 
config.
- Added config-default.mk. Invoking make config adds config-local.mk which 
overwrites any variables from the default config.
- Updated the INSTALL file to include instructions on how to build PaintWeb.
- Made it easier to build PaintWeb by having only a single file to edit paths: 
the config-local.mk file. Until now you had to edit files in the scripts/ 
folder.
- Removed the make moodle target.
- Added the make moodle19 and moodle20 targets, as requested by the Moodle 
developers. The Moodle 2.0 package does not include 1.9-only scripts and 
vice-versa.
- Updated the README.

2009-10-07:
===========

- Using $CFG->libdir in Moodle integration code, as suggested by Martin 
Langhoff. Thanks!
- Fixed the PaintWeb icon in the TinyMCE plugin. Thanks to Mauno Korpelainen for 
his suggestion!

2009-09-11:
===========

- Another fix for the overlay button in the TinyMCE plugin. This time, the fix 
is for Webkit. Yesterday's change did include a regression in Webkit-based 
browsers: the overlay button was not showing up at all.

2009-09-10:
===========

- Made some improvements to the TinyMCE plugin, with regards to the overlay 
button.
- Worked around a problem which caused the overlay button to no longer show up 
in Opera 10 final. The Opera bug is related to losing image selection, due to 
the use of queryCommandState, DOM changes, and mouse events. See Opera bug 
DSK-265135 and TC: http://www.robodesign.ro/coding/0064/

2009-08-27:
===========

- Fixed a Text tool drawing bug which affected Gecko-based browsers. See 
https://bugzilla.mozilla.org/show_bug.cgi?id=478445

2009-08-26:
===========

- Fixed a performance regression on the OLPC XO. The new checkers background 
used for making transparency obvious ... also causes a performance decrease. The 
solution is to disable the background image on the OLPC XO.

- Added a new configuration option to allow enabling/disabling of the checkers 
background image.

2009-08-24:
===========

- Fixed the simple demo to "work" in MSIE. That is: it should nicely tell the 
user that PaintWeb does not work. Before this update, the scripts had JS 
"errors".

- Improved the Bézier curve. When only two points were available, the line drawn 
didn't use the correct color.

src/tools/text.js
/*
 * Copyright (c) 2009-2014, Mihai Sucan
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * 3. Neither the name of the copyright holder nor the names of its contributors
 *    may be used to endorse or promote products derived from this software without
 *    specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * $URL: http://code.google.com/p/paintweb $
 * $Date: 2014-01-28 13:07:03 $
 */

/**
 * @author <a lang="ro" href="http://www.robodesign.ro/mihai">Mihai Şucan</a>
 * @fileOverview Holds the text tool implementation.
 */

// TODO: make this tool nicer to use.

/**
 * @class The text tool.
 *
 * @param {PaintWeb} app Reference to the main paint application object.
 */
pwlib.tools.text = function (app) {
  var _self         = this,
      clearInterval = app.win.clearInterval,
      config        = app.config.text,
      context       = app.buffer.context,
      doc           = app.doc,
      gui           = app.gui,
      image         = app.image,
      lang          = app.lang,
      MathRound     = Math.round,
      mouse         = app.mouse,
      setInterval   = app.win.setInterval;

  /**
   * The interval ID used for invoking the drawing operation every few 
   * milliseconds.
   *
   * @private
   * @see PaintWeb.config.toolDrawDelay
   */
  var timer = null;

  /**
   * Holds the previous tool ID.
   *
   * @private
   * @type String
   */
  var prevTool = app.tool ? app.tool._id : null;

  /**
   * Tells if the drawing canvas needs to be updated or not.
   *
   * @private
   * @type Boolean
   * @default false
   */
  var needsRedraw = false;

  var inputString = null,
      input_fontFamily = null,
      ev_configChangeId = null,
      ns_svg = "http://www.w3.org/2000/svg",
      svgDoc = null,
      svgText = null,
      textWidth = 0,
      textHeight = 0;

  /**
   * Tool preactivation code. This method check if the browser has support for 
   * rendering text in Canvas.
   *
   * @returns {Boolean} True if the tool can be activated successfully, or false 
   * if not.
   */
  this.preActivate = function () {
    if (!gui.inputs.textString || !gui.inputs.text_fontFamily || 
        !gui.elems.viewport) {
      return false;

    }

    // Canvas 2D Text API
    if (context.fillText && context.strokeText) {
      return true;
    }

    // Opera can only render text via SVG Text.
    // Note: support for Opera has been disabled.
    // There are severe SVG redraw issues when updating the SVG text element.
    // Besides, there are important memory leaks.
    // Ultimately, there's a deal breaker: security violation. The SVG document 
    // which is rendered inside Canvas is considered "external" 
    // - get/putImageData() and toDataURL() stop working after drawImage(svg) is 
    // invoked. Eh.
    /*if (pwlib.browser.opera) {
      return true;
    }*/

    // Gecko 1.9.0 had its own proprietary Canvas 2D Text API.
    if (context.mozPathText) {
      return true;
    }

    alert(lang.errorTextUnsupported);
    return false;
  };

  /**
   * The tool activation code. This sets up a few variables, starts the drawing 
   * timer and adds event listeners as needed.
   */
  this.activate = function () {
    // Reset the mouse coordinates to the scroll top/left corner such that the 
    // text is rendered there.
    mouse.x = Math.round(gui.elems.viewport.scrollLeft / image.canvasScale),
    mouse.y = Math.round(gui.elems.viewport.scrollTop  / image.canvasScale),

    input_fontFamily = gui.inputs.text_fontFamily;
    inputString = gui.inputs.textString;

    if (!context.fillText && pwlib.browser.opera) {
      ev_configChangeId = app.events.add('configChange', ev_configChange_opera);
      inputString.addEventListener('input',  ev_configChange_opera, false);
      inputString.addEventListener('change', ev_configChange_opera, false);
    } else {
      ev_configChangeId = app.events.add('configChange', ev_configChange);
      inputString.addEventListener('input',  ev_configChange, false);
      inputString.addEventListener('change', ev_configChange, false);
    }

    // Render text using the Canvas 2D context text API defined by HTML 5.
    if (context.fillText && context.strokeText) {
      _self.draw = _self.draw_spec;

    } else if (pwlib.browser.opera) {
      // Render text using a SVG Text element which is copied into Canvas using 
      // drawImage().
      _self.draw = _self.draw_opera;
      initOpera();

    } else if (context.mozPathText) {
      // Render text using proprietary API available in Gecko 1.9.0.
      _self.draw = _self.draw_moz;
      textWidth = context.mozMeasureText(inputString.value);
    }

    if (!timer) {
      timer = setInterval(_self.draw, app.config.toolDrawDelay);
    }
    needsRedraw = true;
  };

  /**
   * The tool deactivation simply consists of removing the event listeners added 
   * when the tool was constructed, and clearing the buffer canvas.
   */
  this.deactivate = function () {
    if (timer) {
      clearInterval(timer);
      timer = null;
    }
    needsRedraw = false;

    if (ev_configChangeId) {
      app.events.remove('configChange', ev_configChangeId);
    }

    if (!context.fillText && pwlib.browser.opera) {
      inputString.removeEventListener('input',  ev_configChange_opera, false);
      inputString.removeEventListener('change', ev_configChange_opera, false);
    } else {
      inputString.removeEventListener('input',  ev_configChange, false);
      inputString.removeEventListener('change', ev_configChange, false);
    }

    svgText = null;
    svgDoc = null;

    context.clearRect(0, 0, image.width, image.height);

    return true;
  };

  /**
   * Initialize the SVG document for Opera. This is used for rendering the text.
   * @private
   */
  function initOpera () {
    svgDoc = doc.createElementNS(ns_svg, 'svg');
    svgDoc.setAttributeNS(ns_svg, 'version', '1.1');

    svgText = doc.createElementNS(ns_svg, 'text');
    svgText.appendChild(doc.createTextNode(inputString.value));
    svgDoc.appendChild(svgText);

    svgText.style.font = context.font;

    if (app.config.shapeType !== 'stroke') {
      svgText.style.fill = context.fillStyle;
    } else {
      svgText.style.fill = 'none';
    }

    if (app.config.shapeType !== 'fill') {
      svgText.style.stroke = context.strokeStyle;
      svgText.style.strokeWidth = context.lineWidth;
    } else {
      svgText.style.stroke = 'none';
      svgText.style.strokeWidth = context.lineWidth;
    }

    textWidth  = svgText.getComputedTextLength();
    textHeight = svgText.getBBox().height;

    svgDoc.setAttributeNS(ns_svg, 'width',  textWidth);
    svgDoc.setAttributeNS(ns_svg, 'height', textHeight + 10);
    svgText.setAttributeNS(ns_svg, 'x', 0);
    svgText.setAttributeNS(ns_svg, 'y', textHeight);
  };

  /**
   * The <code>configChange</code> application event handler. This is also the 
   * <code>input</code> and <code>change</code> event handler for the text 
   * string input element.  This method updates the Canvas text-related 
   * properties as needed, and re-renders the text.
   *
   * <p>This function is not used on Opera.
   *
   * @param {Event|pwlib.appEvent.configChange} ev The application/DOM event 
   * object.
   */
  function ev_configChange (ev) {
    if (ev.type === 'input' || ev.type === 'change' ||
        (!ev.group && ev.config === 'shapeType') ||
        (ev.group === 'line' && ev.config === 'lineWidth')) {
      needsRedraw = true;

      // Update the text width.
      if (!context.fillText && context.mozMeasureText) {
        textWidth = context.mozMeasureText(inputString.value);
      }
      return;
    }

    if (ev.type !== 'configChange' && ev.group !== 'text') {
      return;
    }

    var font = '';

    switch (ev.config) {
      case 'fontFamily':
        if (ev.value === '+') {
          fontFamilyAdd(ev);
        }
      case 'bold':
      case 'italic':
      case 'fontSize':
        if (config.bold) {
          font += 'bold ';
        }
        if (config.italic) {
          font += 'italic ';
        }
        font += config.fontSize + 'px ' + config.fontFamily;
        context.font = font;

        if ('mozTextStyle' in context) {
          context.mozTextStyle = font;
        }

      case 'textAlign':
      case 'textBaseline':
        needsRedraw = true;
    }

    // Update the text width.
    if (ev.config !== 'textAlign' && ev.config !== 'textBaseline' && 
        !context.fillText && context.mozMeasureText) {
      textWidth = context.mozMeasureText(inputString.value);
    }
  };

  /**
   * The <code>configChange</code> application event handler. This is also the 
   * <code>input</code> and <code>change</code> event handler for the text 
   * string input element.  This method updates the Canvas text-related 
   * properties as needed, and re-renders the text.
   *
   * <p>This is function is specific to Opera.
   *
   * @param {Event|pwlib.appEvent.configChange} ev The application/DOM event 
   * object.
   */
  function ev_configChange_opera (ev) {
    if (ev.type === 'input' || ev.type === 'change') {
      svgText.replaceChild(doc.createTextNode(this.value), svgText.firstChild);
      needsRedraw = true;
    }

    if (!ev.group && ev.config === 'shapeType') {
      if (ev.value !== 'stroke') {
        svgText.style.fill = context.fillStyle;
      } else {
        svgText.style.fill = 'none';
      }

      if (ev.value !== 'fill') {
        svgText.style.stroke = context.strokeStyle;
        svgText.style.strokeWidth = context.lineWidth;
      } else {
        svgText.style.stroke = 'none';
        svgText.style.strokeWidth = context.lineWidth;
      }
      needsRedraw = true;
    }

    if (!ev.group && ev.config === 'fillStyle') {
      if (app.config.shapeType !== 'stroke') {
        svgText.style.fill = ev.value;
        needsRedraw = true;
      }
    }

    if ((!ev.group && ev.config === 'strokeStyle') ||
        (ev.group === 'line' && ev.config === 'lineWidth')) {
      if (app.config.shapeType !== 'fill') {
        svgText.style.stroke = context.strokeStyle;
        svgText.style.strokeWidth = context.lineWidth;
        needsRedraw = true;
      }
    }

    if (ev.type === 'configChange' && ev.group === 'text') {
      var font = '';
      switch (ev.config) {
        case 'fontFamily':
          if (ev.value === '+') {
            fontFamilyAdd(ev);
          }
        case 'bold':
        case 'italic':
        case 'fontSize':
          if (config.bold) {
            font += 'bold ';
          }
          if (config.italic) {
            font += 'italic ';
          }
          font += config.fontSize + 'px ' + config.fontFamily;
          context.font = font;
          svgText.style.font = font;

        case 'textAlign':
        case 'textBaseline':
          needsRedraw = true;
      }
    }

    textWidth  = svgText.getComputedTextLength();
    textHeight = svgText.getBBox().height;

    svgDoc.setAttributeNS(ns_svg, 'width',  textWidth);
    svgDoc.setAttributeNS(ns_svg, 'height', textHeight + 10);
    svgText.setAttributeNS(ns_svg, 'x', 0);
    svgText.setAttributeNS(ns_svg, 'y', textHeight);
  };

  /**
   * Add a new font family into the font family drop down. This function is 
   * invoked by the <code>ev_configChange()</code> function when the user 
   * attempts to add a new font family.
   *
   * @private
   *
   * @param {pwlib.appEvent.configChange} ev The application event object.
   */
  function fontFamilyAdd (ev) {
    var new_font = prompt(lang.promptTextFont) || '';
    new_font = new_font.replace(/^\s+/, '').replace(/\s+$/, '') || 
      ev.previousValue;

    // Check if the font name is already in the list.
    var opt, new_font2 = new_font.toLowerCase(),
        n = input_fontFamily.options.length;

    for (var i = 0; i < n; i++) {
      opt = input_fontFamily.options[i];
      if (opt.value.toLowerCase() == new_font2) {
        config.fontFamily = opt.value;
        input_fontFamily.selectedIndex = i;
        input_fontFamily.value = config.fontFamily;
        ev.value = config.fontFamily;

        return;
      }
    }

    // Add the new font.
    opt = doc.createElement('option');
    opt.value = new_font;
    opt.appendChild(doc.createTextNode(new_font));
    input_fontFamily.insertBefore(opt, input_fontFamily.options[n-1]);
    input_fontFamily.selectedIndex = n-1;
    input_fontFamily.value = new_font;
    ev.value = new_font;
    config.fontFamily = new_font;
  };

  /**
   * The <code>mousemove</code> event handler.
   */
  this.mousemove = function () {
    needsRedraw = true;
  };

  /**
   * Perform the drawing operation using standard 2D context methods.
   *
   * @see PaintWeb.config.toolDrawDelay
   */
  this.draw_spec = function () {
    if (!needsRedraw) {
      return;
    }

    context.clearRect(0, 0, image.width, image.height);

    if (app.config.shapeType != 'stroke') {
      context.fillText(inputString.value, mouse.x, mouse.y);
    }

    if (app.config.shapeType != 'fill') {
      context.beginPath();
      context.strokeText(inputString.value, mouse.x, mouse.y);
      context.closePath();
    }

    needsRedraw = false;
  };

  /**
   * Perform the drawing operation in Gecko 1.9.0.
   */
  this.draw_moz = function () {
    if (!needsRedraw) {
      return;
    }

    context.clearRect(0, 0, image.width, image.height);

    var x = mouse.x,
        y = mouse.y;

    if (config.textAlign === 'center') {
      x -= MathRound(textWidth / 2);
    } else if (config.textAlign === 'right') {
      x -= textWidth;
    }

    if (config.textBaseline === 'top') {
      y += config.fontSize;
    } else if (config.textBaseline === 'middle') {
      y += MathRound(config.fontSize / 2);
    }

    context.setTransform(1, 0, 0, 1, x, y);
    context.beginPath();
    context.mozPathText(inputString.value);

    if (app.config.shapeType != 'stroke') {
      context.fill();
    }

    if (app.config.shapeType != 'fill') {
      context.stroke();
    }
    context.closePath();
    context.setTransform(1, 0, 0, 1, 0, 0);

    needsRedraw = false;
  };

  /**
   * Perform the drawing operation in Opera using SVG.
   */
  this.draw_opera = function () {
    if (!needsRedraw) {
      return;
    }

    context.clearRect(0, 0, image.width, image.height);

    var x = mouse.x,
        y = mouse.y;

    if (config.textAlign === 'center') {
      x -= MathRound(textWidth / 2);
    } else if (config.textAlign === 'right') {
      x -= textWidth;
    }

    if (config.textBaseline === 'bottom') {
      y -= textHeight;
    } else if (config.textBaseline === 'middle') {
      y -= MathRound(textHeight / 2);
    }

    context.drawImage(svgDoc, x, y);

    needsRedraw = false;
  };

  /**
   * The <code>click</code> event handler. This method completes the drawing 
   * operation by inserting the text into the layer canvas.
   */
  this.click = function () {
    _self.draw();
    app.layerUpdate();
  };

  /**
   * The <code>keydown</code> event handler allows users to press the 
   * <kbd>Escape</kbd> key to cancel the drawing operation and return to the 
   * previous tool.
   *
   * @param {Event} ev The DOM Event object.
   * @returns {Boolean} True if the key was recognized, or false if not.
   */
  this.keydown = function (ev) {
    if (!prevTool || ev.kid_ != 'Escape') {
      return false;
    }

    mouse.buttonDown = false;
    app.toolActivate(prevTool, ev);

    return true;
  };
};

// vim:set spell spl=en fo=wan1croqlt tw=80 ts=2 sw=2 sts=2 sta et ai cin fenc=utf-8 ff=unix:

